test/features/bootstrapper/remote_image_download_csi.go

```diff
@@ -0,0 +1,101 @@
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/e2e-framework/klient/k8s"
+	"sigs.k8s.io/e2e-framework/pkg/envconf"
+	"sigs.k8s.io/e2e-framework/pkg/features"
+)
+
+const (
+	bootstrapperImage = "quay.io/dynatrace/dynatrace-bootstrapper:snapshot"
+)
+
+func InstallWithCSI(t *testing.T) features.Feature {
+	builder := features.New("remote-image-download-with-csi")
+	secretConfig := tenant.GetSingleTenantSecret(t)
+
+	appMonDynakube := *dynakubeComponents.New(
+		dynakubeComponents.WithName("app-codemodules"),
+		dynakubeComponents.WithApplicationMonitoringSpec(&oneagent.ApplicationMonitoringSpec{AppInjectionSpec: oneagent.AppInjectionSpec{CodeModulesImage: bootstrapperImage}}),
+		dynakubeComponents.WithAnnotations(map[string]string{dynakube.AnnotationFeatureRemoteImageDownload: "true"}),
+		dynakubeComponents.WithApiUrl(secretConfig.ApiUrl),
+	)
+
+	sampleNamespace := *namespace.New("codemodules-sample-remote-image-download")
+	sampleApp := sample.NewApp(t, &appMonDynakube,
+		sample.AsDeployment(),
+		sample.WithNamespace(sampleNamespace),
+	)
+
+	builder.Assess("create sample namespace", sampleApp.InstallNamespace())
+
+	dynakubeComponents.Install(builder, helpers.LevelAssess, &secretConfig, appMonDynakube)
+
+	builder.Assess("check if jobs completed", jobsAreCompleted(appMonDynakube))
+
+	builder.Assess("check if jobs got cleaned up", jobsAreCleanedUp(appMonDynakube))
+
+	builder.Assess("install sample app", sampleApp.Install())
+
+	builder.Assess("codemodules have been downloaded", codemodules.ImageHasBeenDownloaded(appMonDynakube))
+	builder.Assess("volumes are mounted correctly", codemodules.VolumesAreMountedCorrectly(*sampleApp))
+
+	builder.Teardown(sampleApp.Uninstall())
+	dynakubeComponents.Delete(builder, helpers.LevelTeardown, appMonDynakube)
+
+	return builder.Feature()
+}
+
+func jobsAreCompleted(dk dynakube.DynaKube) features.Func {
+	return func(ctx context.Context, t *testing.T, envConfig *envconf.Config) context.Context {
+		resource := envConfig.Client().Resources()
+
+		jobList := job.GetJobsForOwner(ctx, t, resource, dk.Name, dk.Namespace)
+
+		for _, job := range jobList.Items {
+			t.Logf("waiting for job to be completed: %s", job.Name)
+			pods := pod.GetPodsForOwner(ctx, t, resource, job.Name, job.Namespace)
+			for _, currentPod := range pods.Items {
+				// let's wait not only for the pods to be deleted but rather exactly check if they really reached the state Completed
+				ctx = pod.WaitForCondition(currentPod.Name, currentPod.Namespace, func(object k8s.Object) bool {
+					p, isPod := object.(*corev1.Pod)
+
+					return isPod && len(p.Status.ContainerStatuses) > 0 && p.Status.ContainerStatuses[0].State.Terminated != nil && p.Status.ContainerStatuses[0].State.Terminated.Reason == "Completed"
+				}, time.Minute*5)(ctx, t, envConfig)
+
+				ctx = pod.WaitForPodsDeletionWithOwner(job.Name, job.Namespace)(ctx, t, envConfig)
+			}
+		}
```

Comment:
Is this not flacky?

asking because the `provisioner` will just shoot down the `Job` if it sees the folder it wants to see
so it may happen that you do not see Complete but that is not necessarily a problem

also if you accidentally run this test in an istio env, which would only happen in "local" testing, you will never see Complete (I don't know if which container will our container be (I think istio-proxy is the first))

Reply:
I was also thinking about only waiting for the pods to be deleted, but I thought that may not be precise enough. WDYT does this suffice? To wait only for the pods to be deleted?

Reply:
I think it is enough to check that they appear, and that they go away.
and a sample app gets injected successfully

Reply:
done https://github.com/Dynatrace/dynatrace-operator/pull/4599/commits/90ef1349e62426f2ca489b64ae614d460dad451e